Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/widget/CancellableToast.kt

```diff
@@ -0,0 +1,61 @@
+package com.github.adamantcheese.chan.ui.widget
```

Comment:
I'm not entirely sure of the purpose of this class? If you display a toast while one is already being displayed, the previous one will be automatically hidden by the Toast's internal methods. You can see that code here https://android.googlesource.com/platform/frameworks/base/+/f4bed684c939b0f8809ef404b8609fe4ef849263/core/java/android/widget/Toast.java#426

Reply:
You can also test it for yourself by putting this code snippet into the onCreate method in StartActivity. ```BackgroundUtils.runOnUiThread(() -> BackgroundUtils.runWithExecutor(Executors.newSingleThreadExecutor(), () -> {
            Random r = new Random();
            while (true) {
                showToast(String.valueOf(r.nextInt()));
                long time = System.currentTimeMillis();
                while (System.currentTimeMillis() - time < 1000) {}
            }
        }), 10000);```

Reply:
Well it isn't removed. It's just queued up with other toasts and they all are being shown at the same time.

Reply:
I guess the code in AOSP is correct but only for a single instance of Toast. Since Toast.makeText() returns a new instance every single time you call it, the precious one won't get canceled. And that's exactly why I implemented CancelableToast.

Reply:
If you display a toast, and then immediately display another toast, what happens is that the first toast's view is detached from the window and then it is allowed for its timer to run out, while the second toast is displayed. So you'd prefer that, when you display a toast and then immediately display a second toast, that the first toast doesn't sit around and wait to be removed? There's barely any performance impact, and as far as I can tell, it's perfectly fine to spam Toast instances, as there's built-in protection for this kind of situation anyways https://android.googlesource.com/platform/frameworks/base/+/9a7debe5857ffc7c71cfc4082b1b6d72a5cf81cd/services/java/com/android/server/NotificationManagerService.java#694

Reply:
I should probably record a video and compare the two, because there is a noticeable difference.

Reply:
Alright, here is how cancelable toast works: https://files.catbox.moe/szd4fr.mp4
and here is a regular toast: https://files.catbox.moe/l68ee6.mp4
The difference is pretty obvious.

Reply:
Yeah, it is better I agree. Although now I'm wondering if we even need to bother with displaying toasts while doing a batch download? Maybe make that whole "Doing heavy lifting" dialog be a Snackbar instance instead and allow the user can cancel the download from it? You shouldn't need to block the UI while the download is going on in the background.

Reply:
Well, we will have to move the ImageSaver into a separate foreground service then because user may assume that the download should continue to work even if he closes the app.

Reply:
>Maybe make that whole "Doing heavy lifting" dialog be a Snackbar instance instead and allow the user can cancel the download from it? 

Oh yeah, and it's already possible to cancel it via notification.

Reply:
But maybe I should also add a cancel button to the LoadingControllerView?